### ISSUE TITLE
Lint templated file before evaluation

### ISSUE THREAD

**Issue by anna-azizian** [sqlfluff/sqlfluff#3496]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### Description

Lint the actual templated files instead of evaluating `if`s and `for`s.

### Use case

Currently, not all errors are found in the following SQL
```sql
SELECT *
FROM T
{% if some_jinja_condition %}
    where some_condition
{% else %}
    where some_other_condition
{%endif%}
```
If keywords are enforced to be uppercase, the linter finds only one of `where`s depending on how will the `some_jinja_condition` be evaluated. Instead, it could scan the whole document without looking at those templates (for some rules). Thus, the expected result would be 2 errors suggesting to make `where` uppercase.

### Dialect

N/A

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by barrywhart** [comment]
The only way to do this would be to lint the file multiple times with different settings.

**Comment by barrywhart** [comment]
@anna-azizian: Do you use the Jinja templater or the dbt templater? Eventually, it would be good to support this for both templaters, but Jinja would probably be easier to support first.

**Comment by barrywhart** [comment]
Also, if you can provide an example that looks a little more like real code, that would be helpful. Not absolutely necessary, but it may help us understand the expected usage better.

Do you expect to need this for both Jinja variables and Jinja macros? (Your example uses a variable.)

**Comment by anna-azizian** [comment]
@barrywhart Currently I'm using dbt templater, but having jinja support first would be fine.
Here is a code example with macro usage (which is similar to real-world example). Also, L003 suggests a wrong indentation for GROUP BY line, which doesn't happen without the template section.

```sql
SELECT
    a,
    COUNT(*) AS cnt
FROM tbl
WHERE first_condition
    {% if is_incremental() %}
        AND second_condition
    {% endif %}
GROUP BY a
```

**Comment by alanmcruickshank** [comment]
I'm going to attack L003 as part of #3673 which might address solving some of this.

**Comment by alanmcruickshank** [comment]
I'm going to close this issue. I think the templated indents are solved as part of #3673. The part about linting unrendered sections of code isn't yet, but I think that's covered as part of #4061 .
